exonum-java-binding/core/rust/src/storage/proof_map_index.rs

```diff
@@ -207,13 +207,23 @@ fn convert_to_java_proof<'a>(
     proof: MapProof<Key, Value>,
 ) -> JniResult<JObject<'a>> {
     let proof_nodes: JObject = create_java_proof_nodes(&env, &proof)?;
-    let missing_keys: JObject = create_java_missing_keys(&env, &proof)?;
 
-    // TODO: avoid checking proofs (ECR-1802) and reorder the surrounding operations
-    let checked_proof = proof.check().unwrap();
-    let map_entries: JObject = create_java_map_entries(&env, &checked_proof)?;
+    let native_entries: Vec<_> = proof.all_entries_unchecked().collect();
 
-    create_java_unchecked_flat_map_proof(&env, proof_nodes, map_entries, missing_keys)
+    let mut missing_keys = Vec::with_capacity(native_entries.len());
```

Comment:
Please refactor so that this code is easier to grasp. Currently it operates on various level of abstractions (`create_java_proof_nodes` does everything in one go, then there is a loop that does something).

I'd suggest making `create_java_X` below accept a proof (or native_entries in case of performance considerations) so that this code looks
```
create_java_proof_nodes
create_java_missing_keys
create_java_map_entries
create_java_unchecked_map_proof
```

Reply:
🤔 Passing `native_entries` to `create_java_missing_keys` and `create_java_map_entries` will lead to two iterations over `native_entries` instead of one. Not a huge concern, but I want to make this code reasonably effective. Do you think it's premature optimization?

Reply:
It is good ­— the difference is dwarfed by JNI calls cost :)